### PR TITLE
Samples second pass for Clustering Trainer

### PR DESCRIPTION
**Pull request by artidoro**
Tracked in #2522.

This PR removes the dependency on SampleUtils for clustering trainers samples (KMeans samples).

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            var random = new Random(seed);
+            float randomFloat() => (float)random.NextDouble();
+            for (int i = 0; i < count; i++)
+            {
+                int label = i < count / 2 ? 0 : 1;
+                yield return new DataPoint
+                {
+                    Label = (uint)label,
+                    // Create random features with two clusters.
+                    // The first half has feature values cetered around 0.6 the second half has values centered around 0.4.
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: >cetered [](start = 57, length = 7)

centered

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeans.cs [hunk 2]
```diff
@@ -1,51 +1,114 @@
...
-            var centroid = centroids[0].GetValues();
-            Console.WriteLine($"The coordinates of centroid 0 are: ({string.Join(", ", centroid.ToArray())})");
-
-            //  Expected output similar to:
-            //      The coordinates of centroid 0 are: (26, 6, 1)
+            // Expected output similar to:
+            //   The first 3 coordinates of the first centroid are: (0.6035213, 0.6017533, 0.5964218)
+            //   The first 3 coordinates of the second centroid are: (0.4031044, 0.4175443, 0.4082336)
             //
             // Note: use the advanced options constructor to set the number of threads to 1 for a deterministic behavior.
```
Comment (zeahmed) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeans.cs]: >// Note: use the advanced options constructor to set the number of threads to 1 for a deterministic behavior. [](start = 12, length = 109)

This comment seems out of context. Specifically what does advance options constructor mean here?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            var transformedTestData = model.Transform(testData);
+
+            // Convert IDataView object to a list.
+            var predictions = mlContext.Data.CreateEnumerable<Prediction>(transformedTestData, reuseRowObject: false).ToList();
+
+            // Look at 5 predictions
+            foreach (var p in predictions.Take(2))
+                Console.WriteLine($"Label: {p.Label}, Prediction: {p.PredictedLabel}");
+            foreach (var p in predictions.TakeLast(3))
+                Console.WriteLine($"Label: {p.Label}, Prediction: {p.PredictedLabel}");
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: Most of the time, clustering doesn't have a label. Maybe we should make one sample for labeled clustering, one for label-free clustering.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+            var modelParams = model.Model;
+            modelParams.GetClusterCentroids(ref centroids, out int k);
+            Console.WriteLine($"The first 3 coordinates of the first centroid are: ({string.Join(", ", centroids[0].GetValues().ToArray().Take(3))})");
+            Console.WriteLine($"The first 3 coordinates of the second centroid are: ({string.Join(", ", centroids[1].GetValues().ToArray().Take(3))})");
+
+            <#=ExpectedCentroidsOutput#>
+        }
+
+        private static IEnumerable<DataPoint> GenerateRandomDataPoints(int count, int seed = 0)
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: >int count [](start = 71, length = 9)

Set this to the default used above.

Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: >int seed = 0 [](start = 82, length = 12)

I'm superstitious about passing a 0 to someone else's random number generator. Maybe 1?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            var random = new Random(seed);
+            float randomFloat() => (float)random.NextDouble();
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: >(float)random.NextDouble() [](start = 35, length = 26)

- 0.5f

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeansWithOptions.tt [hunk 6]
```diff
@@ -0,0 +1,33 @@
+﻿<#@ include file="Clustering.ttinclude"#>
+<#+
+string ClassName = "KMeansWithOptions";
+string Trainer = "KMeans";
+string TrainerOptions = @"KMeansTrainer.Options
+            {
+                NumberOfClusters = 2,
+                MaximumNumberOfIterations = 100,
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeansWithOptions.tt]: >MaximumNumberOfIterations = 100 [](start = 16, length = 31)

I don't think we should suggest tuning this. Mostly we want this option to *not* be hit.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeans.tt [hunk 7]
```diff
@@ -0,0 +1,29 @@
+﻿<#@ include file="Clustering.ttinclude"#>
+<#+
+string ClassName = "KMeans";
+            string Trainer = "KMeans";
+            string TrainerOptions = null;
+            string InlineTrainerOptions = "numberOfClusters: 2";
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeans.tt]: >numberOfClusters: 2 [](start = 43, length = 19)

10

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude (hunk 4 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: We use 0 in all our templates for the method GenerateRandomDataPoints. I don't see what it should be an issue

---
In reply to: [275080461](https://github.com/dotnet/machinelearning/pull/3317#discussion_r275080461) [](ancestors = 275080461)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude (hunk 5 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: I don't think that centering at 0 will give any significant advantage. 

---
In reply to: [275080652](https://github.com/dotnet/machinelearning/pull/3317#discussion_r275080652) [](ancestors = 275080652)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude (hunk 3 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/Clustering.ttinclude]: I added some comments to explain that the Label column is not used during training. This is simply for comparison with the predicted label.

---
In reply to: [275080292](https://github.com/dotnet/machinelearning/pull/3317#discussion_r275080292) [](ancestors = 275080292)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeans.tt (hunk 7 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Clustering/KMeans.tt]: Are you suggesting to use more clusters?

---
In reply to: [275080967](https://github.com/dotnet/machinelearning/pull/3317#discussion_r275080967) [](ancestors = 275080967)

**Pull request merged**